truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java

```diff
@@ -524,9 +525,65 @@ public final void atomic(Runnable closure) {
     }
 
     /**
+     * Returns <code>true</code> if this node should be considered tagged by a given tag else
+     * <code>false</code>. The method is only invoked for tags which are explicitly declared as
+     * {@link com.oracle.truffle.api.instrumentation.ProvidedTags provided} by the
+     * {@link TruffleLanguage language}. If the {@link #getSourceSection() source section} of the
+     * node returns <code>null</code> then this node is considered to be not tagged by any tag.
+     * <p>
+     * Tags are used by guest languages to indicate that a {@link Node node} is a member of a
+     * certain category of nodes. For example a debugger
+     * {@link com.oracle.truffle.api.instrumentation.TruffleInstrument instrument} might require a
+     * guest language to tag all nodes as halt locations that should be considered as such.
+     * <p>
+     * The node implementor may decide how to implement tagging for nodes. The simplest way to
+     * implement tagging using Java types is by overriding the {@link #isTaggedWith(Class)} method.
+     * This example shows how to tag a node subclass and all its subclasses as expression and
+     * statement:
+     *
+     * <pre>
+     * <code>
+     * &#64;{@link Override}
+     * protected boolean isTaggedWith({@link Class} tag) {
+     *    return tag == ExpressionTag.class || tag == StatementTag.class;
+     * }
+     * </code>
+     * </pre>
+     *
+     * <p>
+     * Often it is impossible to just rely on the node's Java type to implement tagging. This
+     * example shows how to use local state to implement tagging for a node.
+     *
+     * <pre>
+     * <code>
+     * private boolean isDebuggerHalt;
+     * ...
+     * &#64;{@link Override}
+     * protected boolean isTaggedWith({@link Class} tag) {
+     *    return tag == Debugger.HaltTag.class && isDebuggerHalt;
+     * }
+     * </code>
+     * </pre>
+     * <p>
+     * The implementation of isTaggedWith method must ensure that its result is stable after the
+     * parent {@link RootNode root node} was wrapped in a {@link CallTarget} using
+     * {@link TruffleRuntime#createCallTarget(RootNode)}. The result is stable if the result of
+     * calling this method for a particular tag remains always the same.
+     *
+     * @param tag the class {@link com.oracle.truffle.api.instrumentation.ProvidedTags provided} by
+     *            the {@link TruffleLanguage language}
+     * @return <code>true</code> if the node should be considered tagged by a tag else
+     *         <code>false</code>.
+     * @since 0.12
+     */
+    protected boolean isTaggedWith(Class<?> tag) {
+        return false;
```

Comment:
IMO, this interface should be structured in one of two ways, either:
- accept only annotation classes and detect annotations on the node class if we think that this is the way it's used most of the time (and isTaggedWith is just a fallback for the odd case that cannot be tied to the node class), or:
- not make any assumptions about the class being an annotation if we think that it's going to be specified with isTaggedWith most of the time anyway.

I favor the second option, because it has a smaller interface, and because it looks like at least R and Ruby will use isTaggedWith anyway.
In that cast it makes sense to introduce a base class/interface which documents a class being designed to be used as a tag: `isTaggedWith(Class<? extends SyntaxTag> tag)`